연락처 관리 앱 [Step 1] 뉴준성, 밤삭

newJunsung opened the pull request:
# 연락처 관리 앱

## 리뷰어
@protocorn93

## 파일구조
- ios-contact-manager-ui
  + Controller
     - ContactListViewController.swift 
  + Model
     - Contact.swift
        - Contact
        - ContactManager
  + View


## Model 설명

### Contact
- **UUID**
이름과 나이가 같은 서로 다른 사람이 연락처에 추가될 경우가 있을 수도 있기 때문에, 데이터 상에서 분류하기 위해 `UUID`를 사용했습니다.

- **`_phoneNumbers`와 `phoneNumbers`**
    1. `_phoneNumbers`의 경우, 하나의 연락처에 같은 전화번호 여러 개가 동시에 존재할 수 없기 판단하여, 중복을 허용하지 않기 위해 `Set` 컬렉션 타입을 사용했습니다.<br>
    2. `phoneNumbers`의 경우, TableView Cell의 하나의 전화번호를 보여주는 과정에서 `Set`의 `first`프로퍼티를 이용할 경우, `Set`의 무작위성 때문에, 매번 앱을 실행할 때마다 다른 전화번호가 나오게 되는 현상이 발생하여 일관된 하나의 전화번호만 보여주기 위해 사용하게 됐습니다.

- **`fileprivate` 접근 제어자를 사용한 이유** 
`Contact`의 메소드들을 `ContactManager`만 사용할 수 있도록 은닉화를 하고 싶었습니다.  `ContactManager` 내부에 `Contact`를 선언하지 않은 이유는, `TableView`에서 새로 연락처를 추가하게 된다면 `Contact` 구조체가 필요하다고 생각했기 때문에 `fileprivate`으로 메소드를 선언하고 `ContactManager`와 `Contact`를 하나의 파일에서 관리하도록 했습니다. 

```swift
struct Contact: Hashable {
    let uuid = UUID()
    private(set) var name: String
    private(set) var age: Int
    private var _phoneNumbers = Set<PhoneNumber>()
    var phoneNumbers: [PhoneNumber] {
        _phoneNumbers.sorted()
    }
    
    init(name: String, age: Int) {
        self.name = name
        self.age = age
    }
    
    fileprivate mutating func changeName(_ name: String) {
        self.name = name
    }
    
    fileprivate mutating func changeAge(_ age: Int) {
        self.age = age
    }
    
    fileprivate mutating func add(_ phoneNumber: PhoneNumber) {
        _phoneNumbers.insert(phoneNumber)
    }
    
    fileprivate mutating func delete(_ phoneNumber: PhoneNumber) {
        _phoneNumbers.remove(phoneNumber)
    }
    
    fileprivate mutating func modify(from existingPhoneNumber: PhoneNumber, to newPhoneNumber: PhoneNumber) {
        _phoneNumbers.remove(existingPhoneNumber)
        _phoneNumbers.insert(newPhoneNumber)
    }
}
```

### ContactManager
- **연락처 목록을 Dictionary 타입으로 관리한 이유**
각 연락처의 `UUID`는 중복될 가능성이 0에 가깝기 때문에, 해당 연락처를 UUID로 빠르게 접근할 수 있습니다. 또한 배열로 관리하게 되면 수정・삭제 시, 동일한 UUID를 찾아야 되기 때문에 연락처가 시간 복잡도가 O(n)이 되지만, Dictionary로 관리하게 되면 O(1)로 연락처에 접근할 수 때문입니다.
```swift
class ContactManager {
    private(set) var contacts = [UUID: Contact]()
    var contacts: [Contact] {
        _contacts.values.sorted { $0.name < $1.name }
    }
    // ...
}
```

## 고민했던 점
앞선 `_phoneNumber` 설명에서의 이유로 타입을 `Set`으로 해주었는데, 결국 일관된 전화번호를 `Table View`에서 보여주기 위해 `array`로 변환하는 과정을 거쳐야 했습니다. 전화의 추가・삭제・수정할 때와 중복 검사 할 때는 `Set`이 유리하지만, 전화번호 목록을 보여주는 경우엔 차라리 `Set` 보다는 `Array` 타입이 유리할 수도 있을 것 같습니다. 따라서 이번 프로젝트에서 연락처의 전화번호들을 `Set`으로 관리하는게 좋을 지, `Array`로 관리하는게 좋을지 궁금합니다.

newJunsung committed:
[Config] 프로젝트 생성

newJunsung committed:
[Chore] MVC 패턴에 맞게 폴더링

bamsak committed:
[Feat] Contact 모델 생성

bamsak committed:
[Refactor] Contact 모델 수정

bamsak committed:
[Feat] Contact의 CRUD 메소드 생성

newJunsung committed:
[Feat] ContactManager를 통한 연락처 목록 CRUD 함수 생성

newJunsung committed:
[Refactor] Contact CRUD 함수에 예외처리 추가

bamsak committed:
[Feat] TableView 구현

newJunsung committed:
[Refactor] TableViewCell 스타일 변경 및 Contact 프로퍼티 수정

newJunsung committed:
[Refactor] contacts 배열의 연산 프로퍼티로 변경

ContactListViewController의 contactList를
ContactManager 클래스에서 관리할 수 있게 연산 프로퍼티로 변경

newJunsung committed:
[Chore] 접근 제어자 변경

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        cell.textLabel?.text = "\(contact.name)(\(contact.age))"
+        cell.detailTextLabel?.text = "\(contact.phoneNumbers.first ?? "연락처를 추가해주세요.")"
+        
+        return cell
```
기본 셀이기에 값을 외부에서 할당해주고 있지만 이후 스탭이나 다른 프로젝트에서 커스텀 셀을 사용하게 된다면 커스텀 셀에게 데이터를 건내주고 셀 안에서 UI를 구성해주는게 좋아요.

bamsak replied:
Cell을 ContactCell이라는 CustomCell로 따로 구현했습니다!

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/Contact.swift:
```diff
@@ -0,0 +1,93 @@
+    let uuid = UUID()
+    private(set) var name: String
+    private(set) var age: Int
+    private var _phoneNumbers = Set<PhoneNumber>()
```
제가 확인한 프로젝트 명세에는 하나의 연락처에 여러개를 전화번호를 저장한다는 내용은 없어보이는데요. 혹시 제가 틀렸다면 명세서 링크를 받아볼 수 있을까요? 명세서에 없다면 저는 불필요한 스팩이지 않나 싶어요.

bamsak replied:
명세에서 내용이 없었지만 뉴준성과 저의 생각으로 한 사람이 여러개의 전화번호를 가질 수도 있다고 생각하여 전화번호를 Collection Type으로 만들어 프로젝트 진행을 했었습니다. 하나의 전화번호만 가지도록 수정했습니다!

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/Contact.swift:
```diff
@@ -0,0 +1,93 @@
+        self.age = age
+    }
+    
+    fileprivate mutating func changeName(_ name: String) {
+        self.name = name
+    }
+    
+    fileprivate mutating func changeAge(_ age: Int) {
+        self.age = age
+    }
+    
+    fileprivate mutating func add(_ phoneNumber: PhoneNumber) {
+        _phoneNumbers.insert(phoneNumber)
+    }
+    
+    fileprivate mutating func delete(_ phoneNumber: PhoneNumber) {
+        _phoneNumbers.remove(phoneNumber)
+    }
+    
+    fileprivate mutating func modify(from existingPhoneNumber: PhoneNumber, to newPhoneNumber: PhoneNumber) {
+        _phoneNumbers.remove(existingPhoneNumber)
+        _phoneNumbers.insert(newPhoneNumber)
+    }
```
저는 값을 변경하는 주체는 `ContactManager`가 되어야 한다고 생각해요. 직접 값을 변경하다보니 불필요한 `fileprivate` 변수가 붙게 되는 것 같아요.

newJunsung replied:
생각해보니 그게 더 좋은 판단인것 같습니다! ContactManager가 Contact를 변경할 수 있도록 했습니다.

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,35 @@
+import UIKit
+
+final class ContactListViewController: UIViewController {
+    private var contactManager = ContactManager()
```
`let` 으로 선언되는게 맞을 것 같아요.

newJunsung replied:
let으로 수정했습니다!

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/Contact.swift:
```diff
@@ -0,0 +1,93 @@
+
+import Foundation
+
+typealias PhoneNumber = String
```
typealias 선언을 해주신 이유가 궁금해요! ㅎㅎ

newJunsung replied:
프로퍼티인 phoneNumber를 String으로 표현하는 것보다는 PhoneNumber로 보는 것이 다른 사람이 코드를 읽고 이해하는데 도움이 될 것 같아서 typealias를 선언했습니다.

protocorn93 commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        let contact = contactManager.contacts[indexPath.row]
+        
+        cell.textLabel?.text = "\(contact.name)(\(contact.age))"
+        cell.detailTextLabel?.text = "\(contact.phoneNumbers.first ?? "연락처를 추가해주세요.")"
```
000-000-0000 포맷팅이 필요하지 않을까요?

newJunsung replied:
오늘 STEP 2에 대한 정보를 확인했는데, 연락처를 새로 추가하거나 수정할 때 포맷에 맞춰서 저장하도록 하는게 더 좋을 것 같다고 생각합니다.

protocorn93 reviewed (changes requested):
두 분 수고하셨어요. 

코멘트에도 언급했듯이, 하나의 연락처에 다수의 전화번호가 있는 것은 프로젝트 명세가 아닌 것으로 보어져요.  그렇기에 저는 Array, Set 이 둘을 모두 사용할 필요는 없어보이긴해요. 

나머지는 코멘트 확인 부탁드릴게요!

newJunsung committed:
[Refactor] Contact, ContactManager 리팩토링

하나의 사람이 여러가지 전화번호를 가질 수 있다는 가정은 이번 프로젝트에서 염두하지 않은 모양이다.
따라서 전화번호를 단일 타입으로 바꿨음.
ContactManager가 Contact를 관리할 수 있도록 수정.

bamsak committed:
[Refactor] ContactCell CustomCell로 수정
